File: GameStream/Services/GameDiscovery.swift
import Foundation

enum DiscoveryMode: String, CaseIterable, Hashable {
    case multiplayer
    case coOp
    case solo
    case quickPlay
    case gamePass

    var title: String {
        switch self {
        case .multiplayer: return "Multiplayer"
        case .coOp: return "Co-op"
        case .solo: return "Solo"
        case .quickPlay: return "Quick play"
        case .gamePass: return "Game Pass"
        }
    }

    var subtitle: String {
        switch self {
        case .multiplayer: return "Jump in with other players"
        case .coOp: return "Play together on the couch or online"
        case .solo: return "Story and single-player sessions"
        case .quickPlay: return "Short sessions you can finish tonight"
        case .gamePass: return "Ready on Xbox Game Pass"
        }
    }

    var shelfTitle: String {
        switch self {
        case .multiplayer: return "Play with others"
        case .coOp: return "Co-op nights"
        case .solo: return "Solo stories"
        case .quickPlay: return "Quick play"
        case .gamePass: return "On Game Pass"
        }
    }
}

extension GameCatalog {
    static func modes(forId id: String) -> Set<DiscoveryMode> {
        switch id.uppercased() {
        case "9NNX1VVR3KNQ": return [.multiplayer, .gamePass]
        case "9NP1P1WFS0LB": return [.multiplayer, .coOp, .gamePass]
        case "BT5P2X999VH2": return [.multiplayer, .quickPlay]
        case "9NXP44L49SHJ": return [.multiplayer, .coOp, .gamePass]
        case "9P2N57MC619K": return [.multiplayer, .coOp, .gamePass]
        case "BQ1TN1T79V9K": return [.multiplayer, .quickPlay]
        case "9N201KQXS5BM": return [.multiplayer, .quickPlay]
        case "9NCJSXWZTP88": return [.solo, .gamePass]
        case "9PJTHRNVH62H": return [.coOp, .gamePass, .solo]
        case "9NKV34XDW014": return [.multiplayer, .coOp]
        case "9NBR2VXT87SJ": return [.solo, .gamePass]
        case "9NFTC552K3GJ": return [.solo, .quickPlay, .gamePass]
        case "9NX6K9HN4F4K": return [.solo, .gamePass]
        case "9N8CD0XZKLP4": return [.solo, .gamePass]
        case "9NB0115C9WNM": return [.coOp, .solo, .gamePass]
        default: return [.gamePass]
        }
    }

    static func modes(for game: CatalogGame) -> Set<DiscoveryMode> {
        modes(forId: game.id)
    }

    static func games(in mode: DiscoveryMode) -> [CatalogGame] {
        games.filter { modes(for: $0).contains(mode) }
    }

    static func discoveryShelves() -> [(String, [CatalogGame])] {
        DiscoveryMode.allCases.compactMap { mode in
            let items = games(in: mode)
            guard items.count >= 2 else { return nil }
            return (mode.shelfTitle, items)
        }
    }
}

File: android/app/src/main/java/com/gamestream/app/GameDiscovery.kt
package com.gamestream.app

enum class DiscoveryMode(val title: String, val shelfTitle: String) {
    MULTIPLAYER("Multiplayer", "Play with others"),
    CO_OP("Co-op", "Co-op nights"),
    SOLO("Solo", "Solo stories"),
    QUICK_PLAY("Quick play", "Quick play"),
    GAME_PASS("Game Pass", "On Game Pass");
}

object GameDiscovery {
    fun modes(id: String): Set<DiscoveryMode> = when (id.uppercase()) {
        "9NNX1VVR3KNQ" -> setOf(DiscoveryMode.MULTIPLAYER, DiscoveryMode.GAME_PASS)
        "9NP1P1WFS0LB" -> setOf(DiscoveryMode.MULTIPLAYER, DiscoveryMode.CO_OP, DiscoveryMode.GAME_PASS)
        "BT5P2X999VH2" -> setOf(DiscoveryMode.MULTIPLAYER, DiscoveryMode.QUICK_PLAY)
        "9NXP44L49SHJ" -> setOf(DiscoveryMode.MULTIPLAYER, DiscoveryMode.CO_OP, DiscoveryMode.GAME_PASS)
        "9P2N57MC619K" -> setOf(DiscoveryMode.MULTIPLAYER, DiscoveryMode.CO_OP, DiscoveryMode.GAME_PASS)
        "BQ1TN1T79V9K" -> setOf(DiscoveryMode.MULTIPLAYER, DiscoveryMode.QUICK_PLAY)
        "9N201KQXS5BM" -> setOf(DiscoveryMode.MULTIPLAYER, DiscoveryMode.QUICK_PLAY)
        "9NCJSXWZTP88" -> setOf(DiscoveryMode.SOLO, DiscoveryMode.GAME_PASS)
        "9PJTHRNVH62H" -> setOf(DiscoveryMode.CO_OP, DiscoveryMode.GAME_PASS, DiscoveryMode.SOLO)
        "9NKV34XDW014" -> setOf(DiscoveryMode.MULTIPLAYER, DiscoveryMode.CO_OP)
        "9NBR2VXT87SJ" -> setOf(DiscoveryMode.SOLO, DiscoveryMode.GAME_PASS)
        "9NFTC552K3GJ" -> setOf(DiscoveryMode.SOLO, DiscoveryMode.QUICK_PLAY, DiscoveryMode.GAME_PASS)
        "9NX6K9HN4F4K" -> setOf(DiscoveryMode.SOLO, DiscoveryMode.GAME_PASS)
        "9N8CD0XZKLP4" -> setOf(DiscoveryMode.SOLO, DiscoveryMode.GAME_PASS)
        "9NB0115C9WNM" -> setOf(DiscoveryMode.CO_OP, DiscoveryMode.SOLO, DiscoveryMode.GAME_PASS)
        else -> setOf(DiscoveryMode.GAME_PASS)
    }

    fun games(mode: DiscoveryMode): List<CatalogGame> =
        GameCatalog.games.filter { mode in modes(it.id) }

    fun shelves(): List<Pair<String, List<CatalogGame>>> =
        DiscoveryMode.entries.mapNotNull { mode ->
            val items = games(mode)
            if (items.size >= 2) mode.shelfTitle to items else null
        }

    fun fromTitle(title: String): DiscoveryMode? =
        DiscoveryMode.entries.firstOrNull { it.title == title }
}
